src/Std/Data/DHashMap/Internal/List/Associative.lean

```diff
@@ -2751,128 +2827,223 @@ theorem alterKey_cons_perm [BEq α] [LawfulBEq α] {k : α} {f : Option (β k) 
-    {l : List ((a : α) × β a)} (p : (a : α) × β a) (hne : p.1 ≠ a) :
-    p ∈ alterKey a f l ↔ p ∈ l := by
+theorem getKey?_alterKey {k k' : α} {f : Option (β k) → Option (β k)} (l : List ((a : α) × β a))
+    (hl : DistinctKeys l) :
+    getKey? k' (alterKey k f l) =
+      if k == k' then
+        if (f (getValueCast? k l)).isSome then some k else none
+      else
+        getKey? k' l := by
   rw [alterKey]
-  split <;> simp only [mem_eraseKey_of_key_ne p hne, mem_insertEntry_of_key_ne p hne]
+  split <;> next heq => simp [hl, heq, getKey?_eraseKey, getKey?_insertEntry]
 
-theorem length_modifyKey [BEq α] [LawfulBEq α] (k : α) (f : β k → β k)
-    (l : List ((a : α) × β a)) : (modifyKey k f l).length = l.length := by
-  induction l
+theorem getKey!_alterKey [Inhabited α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (l : List ((a : α) × β a)) (hl : DistinctKeys l) :
+    getKey! k' (alterKey k f l) =
+      if k == k' then
+        if (f (getValueCast? k l)).isSome then k else panic ""
+      else
+        getKey! k' l := by
+  simp only [getKey!_eq_getKey?, hl, getKey?_alterKey, beq_iff_eq]
+  split
+  · next heq =>
+    cases eq_of_beq heq
+    split <;> rfl
+  · next heq =>
+    rfl
+
+theorem getKey_alterKey [Inhabited α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (l : List ((a : α) × β a)) (hl : DistinctKeys l) (hc : containsKey k' (alterKey k f l)) :
+    getKey k' (alterKey k f l) hc =
+      if heq : k == k' then
+        k
+      else
+        haveI h' : containsKey k' l := by rwa [containsKey_alterKey hl, if_neg heq] at hc
+        getKey k' l h' := by
+  have := getKey?_alterKey (f := f) (k' := k') _ hl
+  rw [getKey?_eq_some_getKey hc] at this
+  split
+  · next heq =>
+    cases eq_of_beq heq
+    apply Option.some_inj.mp
+    simp_all
+  · next heq =>
+    apply Option.some_inj.mp
+    simp_all only [Bool.false_eq_true, ite_false]
+    rw [getKey?_eq_some_getKey]
+
+theorem getKeyD_alterKey {k k' d : α} {f : Option (β k) → Option (β k)} (l : List ((a : α) × β a))
+    (hl : DistinctKeys l) :
+    getKeyD k' (alterKey k f l) d =
+      if k == k' then
+        if (f (getValueCast? k l)).isSome then k else d
+      else
+        getKeyD k' l d := by
+  simp only [hl, getKeyD_eq_getKey?, getKey?_alterKey, beq_iff_eq, Function.comp_apply]
+  split
+  · next heq =>
+    cases eq_of_beq heq
+    split <;> rfl
   · rfl
-  · next ih =>
-    simp only [modifyKey]
-    split <;> next h => simp only [length_replaceEntry, List.length_cons]
-
-theorem containsKey_modifyKey_self [BEq α] [LawfulBEq α] (k : α) (f : β k → β k)
-    (l : List ((a : α) × β a)) : containsKey k (modifyKey k f l) = containsKey k l := by
-  induction l
-  · simp only [modifyKey, getValueCast?_nil, eraseKey_nil, containsKey_nil, Bool.false_eq_true]
-  · simp only [modifyKey, Bool.coe_iff_coe]
-    split
-    · rfl
-    · rw [containsKey_replaceEntry]
 
 namespace Const
 
+omit [LawfulBEq α]
```

Comment:
`omit`, and especially `omit ... in` are supposed to be last-resort constructs. Please reorganize the sectioning so that you don't need them here.